crates/language/src/buffer.rs

```diff
@@ -4145,6 +4145,56 @@ impl BufferSnapshot {
             None
         }
     }
+
+    pub fn words_in_range(&self, query: Option<&str>, range: Range<usize>) -> HashSet<String> {
+        if query.map_or(false, |query| query.is_empty()) {
+            return HashSet::default();
+        }
+
+        let classifier = CharClassifier::new(self.language.clone().map(|language| LanguageScope {
+            language,
+            override_id: None,
+        }));
+
+        let mut query_ix = 0;
+        let query = query.map(|query| query.chars().collect::<Vec<_>>());
+        let query_len = query.as_ref().map_or(0, |query| query.len());
+
+        let mut words = HashSet::default();
+        let mut current_word_start_ix = None;
+        let mut chunk_ix = range.start;
+        for chunk in self.chunks(range, false) {
+            for (i, c) in chunk.text.char_indices() {
+                let ix = chunk_ix + i;
+                if classifier.is_word(c) {
+                    if current_word_start_ix.is_none() {
+                        current_word_start_ix = Some(ix);
+                    }
+                    if query_ix == query_len {
+                        continue;
+                    } else {
+                        match &query {
+                            Some(query) =>  if c.eq_ignore_ascii_case(query.get(query_ix).expect(
+                                "query_ix is updated with query chars' length and should match the indices",
+                            )) {
+                                query_ix += 1;
```

Comment:
Does this work when the query contains multibyte characters?

Reply:
It should, because I've split the query into `[char; N]` and access that array by indices: so, the algorithm here accesses array entry-by-entry instead of accessing the string char-by-char (which will panic indeed).

Good point that I'd better test this, will do, thanks.